search-provider/eks-metadata-provider.c

```diff
@@ -0,0 +1,602 @@
+                                    GParamSpec *pspec)
+{
+  EksMetadataProvider *self = EKS_METADATA_PROVIDER (object);
+
+  switch (prop_id)
+    {
+    case PROP_APPLICATION_ID:
+      g_value_set_string (value, self->application_id);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+eks_metadata_provider_set_property (GObject      *object,
+                                    unsigned      prop_id,
+                                    const GValue *value,
+                                    GParamSpec   *pspec)
+{
+  EksMetadataProvider *self = EKS_METADATA_PROVIDER (object);
+
+  switch (prop_id)
+    {
+    case PROP_APPLICATION_ID:
+      g_clear_pointer (&self->application_id, g_free);
+      self->application_id = g_value_dup_string (value);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+eks_metadata_provider_finalize (GObject *object)
+{
+  EksMetadataProvider *self = EKS_METADATA_PROVIDER (object);
+
+  g_clear_pointer (&self->application_id, g_free);
+  g_clear_object (&self->skeleton);
+  g_clear_pointer (&self->translation_infos, g_hash_table_unref);
+
+  G_OBJECT_CLASS (eks_metadata_provider_parent_class)->finalize (object);
+}
+
+static void
+eks_metadata_provider_class_init (EksMetadataProviderClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->get_property = eks_metadata_provider_get_property;
+  object_class->set_property = eks_metadata_provider_set_property;
+  object_class->finalize = eks_metadata_provider_finalize;
+
+  eks_metadata_provider_props[PROP_APPLICATION_ID] =
+    g_param_spec_string ("application-id", "Application Id", "Application Id",
+      "", G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  g_object_class_install_properties (object_class,
+                                     NPROPS,
+                                     eks_metadata_provider_props);
+}
+
+typedef struct _MetadataQueryState {
+  EksMetadataProvider   *provider;
+  GDBusMethodInvocation *invocation;
+} MetadataQueryState;
+
+static MetadataQueryState *
+metadata_query_state_new (EksMetadataProvider *provider,
+                                  GDBusMethodInvocation      *invocation)
+{
+  MetadataQueryState *state = g_new0 (MetadataQueryState, 1);
+  state->provider = provider;
+  state->invocation = g_object_ref (invocation);
+
+  return state;
+}
+
+static void
+metadata_query_state_free (MetadataQueryState *state)
+{
+  g_clear_object (&state->invocation);
+
+  g_free (state);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (MetadataQueryState,
+                               metadata_query_state_free);
+
+typedef struct GSList ObjectSList;
+
+static void
+object_slist_free (ObjectSList *slist)
+{
+  g_slist_free_full ((GSList *) slist, g_object_unref);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ObjectSList, object_slist_free);
```

Comment:
My previous comment on this still stands. I find it too much indirection for the benefit it provides.

Reply:
Hrm okay. I like autoptr, but it shouldn't be too bad to just clean these two up manually. I'll add a comment.

Reply:
I'm also OK with using `__attribute__((cleanup))` directly, but I find the typedef really impedes readability.